### PR TITLE
[ios][platform_view] Use CAShapeLayer as the mask to avoid software rendering 

### PR DESCRIPTION
**Pull request by hellohuanlin**
This PR uses `CAShapeLayer` as the mask to avoid software rendering. 

I kept `UIView` as the mask, so that we can measure just the improvement related to avoiding software rendering. This also allows me to land this change sooner. I created [a separate issue](https://github.com/flutter/flutter/issues/149212) to track removing UIView as the mask. 

Note: the previous behavior seems to be incorrect (or at least not pixel perfect). This PR fixed it. See comments. 

See design doc: https://docs.google.com/document/d/1TqG_N4GK_qctuk73Gk3zOdAiILUrwMqxoCMgroK_AeA/edit?resourcekey=0-jUiidfzIS642ngG2w9vSUA&tab=t.0

*List which issues are fixed by this PR. You must list at least one issue.*

Fixes https://github.com/flutter/flutter/issues/142813
Fixes https://github.com/flutter/flutter/issues/142830

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm [hunk 1]
```diff
@@ -1741,17 +1741,15 @@ - (void)testClipRect {
...
       CGPoint point = CGPointMake(i, j);
       int alpha = [self alphaOfPoint:CGPointMake(i, j) onView:mockFlutterView];
-      // Edges of the clipping might have a semi transparent pixel, we only check the pixels that
-      // are fully inside the clipped area.
-      CGRect insideClipping = CGRectMake(3, 3, 1, 1);
       if (CGRectContainsPoint(insideClipping, point)) {
         XCTAssertEqual(alpha, 255);
       } else {
-        XCTAssertLessThan(alpha, 255);
+        XCTAssertEqual(alpha, 0);
```
Comment (hellohuanlin) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm:1752]: The previous behavior is incorrect - there shouldn't be any semi-transparent pixels. This PR fixed it.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm [hunk 2]
```diff
@@ -1816,17 +1814,42 @@ - (void)testClipRRect {
...
+  */
+  CGRect innerClipping1 = CGRectMake(3, 2, 4, 6);
+  CGRect innerClipping2 = CGRectMake(2, 3, 6, 4);
+  CGRect outterClipping = CGRectMake(2, 2, 6, 6);
   for (int i = 0; i < 10; i++) {
     for (int j = 0; j < 10; j++) {
       CGPoint point = CGPointMake(i, j);
       int alpha = [self alphaOfPoint:CGPointMake(i, j) onView:mockFlutterView];
-      // Edges of the clipping might have a semi transparent pixel, we only check the pixels that
-      // are fully inside the clipped area.
```
Comment (hellohuanlin) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm:1824]: Again the existing behavior is incorrect. 

In the picture below, the **black rounded rect** is the mask; the **red rect** is `innerClipping1`, and **green rect** is `innerClipping2`. 

Pixels within the red or green rect should be fully opaque; pixels outside of `outterClipping` should be fully transparent; the 4 corner pixels should be partially transparent. 

IMG_7331

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm [hunk 3]
```diff
@@ -1816,17 +1814,42 @@ - (void)testClipRRect {
...
+  2   + - - +           2
+  3   |     |           3 + - - - - +
+  4   |     |           4 |         |
+  5   |     |           5 |         |
+  6   |     |           6 + - - - - +
+  7   + - - +           7
+  */
+  CGRect innerClipping1 = CGRectMake(3, 2, 4, 6);
+  CGRect innerClipping2 = CGRectMake(2, 3, 6, 4);
+  CGRect outterClipping = CGRectMake(2, 2, 6, 6);
```
Comment (hellohuanlin) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm:1838]: Again the existing behavior is incorrect. 

In the picture below, the **black rounded rect** is the mask. 

Pixels within the **red rect** (`innerClipping1`) or **green rect** (`innerClipping2`) should be fully opaque; pixels outside of **black bounding rect** (`outterClipping`) should be fully transparent; the **4 corner pixels** (`(2, 2), (2, 7), (7, 2) and (7, 7)`) should be partially transparent. 

IMG_7331

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm [hunk 4]
```diff
@@ -1741,17 +1741,15 @@ - (void)testClipRect {
...
   [mockFlutterView layoutIfNeeded];
 
+  CGRect insideClipping = CGRectMake(2, 2, 3, 3);
   for (int i = 0; i < 10; i++) {
     for (int j = 0; j < 10; j++) {
       CGPoint point = CGPointMake(i, j);
       int alpha = [self alphaOfPoint:CGPointMake(i, j) onView:mockFlutterView];
-      // Edges of the clipping might have a semi transparent pixel, we only check the pixels that
-      // are fully inside the clipped area.
-      CGRect insideClipping = CGRectMake(3, 3, 1, 1);
```
Comment (hellohuanlin) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm:1750]: The existing behavior is incorrect - there shouldn't be any semi-transparent pixels on the edge.

**Comment by hellohuanlin** [comment]
I am able to confirm the procedures related to software rendering are gone. For example, `CGBlt_fillBytes` and `ripc_render`. 

Screenshot 2024-05-24 at 9 37 05 AM

Screenshot 2024-05-24 at 9 36 50 AM

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm (hunk 4 above)
Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm:1750]: (Introduced in #20050)

**Review by jonahwilliams — APPROVED** [review]
LGTM!

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.mm [hunk 5]
```diff
@@ -271,15 +271,31 @@ - (instancetype)initWithFrame:(CGRect)frame screenScale:(CGFloat)screenScale {
...
-  paths_.clear();
+  CGPathRelease(pathSoFar_);
+  pathSoFar_ = CGPathCreateMutable();
+  [self shapeLayer].path = nil;
   [self setNeedsDisplay];
 }
 
+- (void)dealloc {
+  CGPathRelease(pathSoFar_);
+  [super dealloc];
```
Comment (hellohuanlin) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.mm:296]: @jmagman fyi potentially code conflict with your ARC migration PRs

**Comment by jmagman** [comment]
Unsurprisingly there's a scenario test failure here.  

Expected: [checked in screenshot](https://github.com/flutter/engine/blob/main/testing/scenario_app/ios/Scenarios/ScenariosUITests/golden_platform_view_cliprrect_with_transform_iPhone%20SE%20(3rd%20generation)_17.0_simulator.png)
golden_platform_view_cliprrect_with_transform_iPhone SE (3rd generation)_17 0_simulator

Actual: [from the xctest](https://storage.googleapis.com/flutter_logs/flutter/72feaa0e-a3cc-446f-8bfe-ebc0360469e4/Scenario%20App%20Integration%20Tests/72feaa0e-a3cc-446f-8bfe-ebc0360469e4/ios_scenario_xcresult.zip)
golden_platform_view_cliprrect_with_transform_iPhone SE (3rd generation)_17 0_simulator_actual

**Comment by hellohuanlin** [comment]
This golden failure is a bit strange: 

> 2024-05-29 14:13:09.200565-0700 ScenariosUITests-Runner[95948:735109] GOLDEN DIFF FAILED: image diff greater than threshold. Current diff: 0.7507098090416192, threshold: 0.5
    t =     4.09s Added attachment named 'golden_platform_view_cliprect_with_transform_iPhone SE (3rd generation)_17.0_simulator_actual.png'

I can confirm that I have regenerated this image from my simulator iPhone SE (3rd gen) 17.0.

**Comment by jmagman** [comment]
> I can confirm that I have regenerated this image from my simulator iPhone SE (3rd gen) 17.0.

Maybe something super subtle with the macOS version? Can you pull the attachment right off the failing test or use the one I attached from the failing test?

**Comment by hellohuanlin** [comment]
@cbracken friendly ping since you requested review, but no rush :)

**Review by cbracken — APPROVED** [review]
This looks great; I'm assuming this class is never accessed from multiple threads? Otherwise consider allocating the new mutable path to a temp, freeing the old one, and then setting the ivar to the temp.

LGTM stamp from a Japanese personal seal

**Comment by hellohuanlin** [comment]
> I'm assuming this class is never accessed from multiple threads? 

Yes, only on main thread.

**Comment by jason-simmons** [comment]
Reason for revert: compilation errors on iOS targets

```
FlutterPlatformViews_Internal.mm:277:10: error: ARC forbids explicit message send of 'dealloc'
  277 |   [super dealloc];
      |    ~~~~~ ^
 ```

(see https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8746006425774972161/+/u/build_ci_ios_debug_unopt_sim_flutter_testing_scenario_app_flutter_shell_platform_darwin_ios:ios_test_flutter/stdout)

**Pull request merged**